packages/typespec-azure-portal-core/src/decorators.ts

```diff
@@ -0,0 +1,254 @@
+import { ArmResourceKind, getArmResourceKind } from "@azure-tools/typespec-azure-resource-manager";
+import {
+  CompilerHost,
+  DecoratorContext,
+  Model,
+  ModelProperty,
+  Program,
+  StringLiteral,
+  Type,
+  getDirectoryPath,
+  getSourceLocation,
+  normalizePath,
+  resolvePath,
+  validateDecoratorUniqueOnNode,
+} from "@typespec/compiler";
+import { PortalCoreKeys } from "./keys.js";
+import { reportDiagnostic } from "./lib.js";
+import { AboutOptions, BrowseOptions, marketplaceOfferOptions } from "./types.js";
+
+/**
+ * This is a Browse decorator which will be use to put more info on the browse view.
+ * @param target The model that is being decorated.
+ * @param options BrowseOptions of the property.
+ */
+export function $browse(context: DecoratorContext, target: Model, options: Model) {
+  const { program } = context;
+  validateDecoratorUniqueOnNode(context, target, $browse);
+  isARMResource(program, target, "browse");
+  const browseOptionsResult: BrowseOptions = {};
+  if (options && options.properties) {
+    const query = options.properties.get("argQuery");
+    if (query && query.type.kind === "Model") {
+      //use decoratorTarget to find sourceLocation instead of target, since we want to find a file path related to where decorator was stated
+      const decoratorTarget = context.decoratorTarget;
+      const sourceLocation = getSourceLocation(decoratorTarget);
+      const dirPath =
+        sourceLocation && sourceLocation.file.path && getDirectoryPath(sourceLocation.file.path);
+      const argQueryPath = query.type && (query.type as Model).properties.get("filePath");
+      const argQueryPathValue =
+        argQueryPath && argQueryPath.type && (argQueryPath.type as StringLiteral).value;
+      let filePath = resolvePath(dirPath, argQueryPathValue); //if given path is fullpath, it will return the fullPath
+      if (filePath && argQueryPath && argQueryPathValue) {
+        filePath = normalizePath(filePath);
+        browseOptionsResult.argQuery = {
+          filePath: filePath,
+        };
+        program.stateMap(PortalCoreKeys.browse).set(target, browseOptionsResult);
+      }
+    } else if (query?.type.kind === "String") {
+      browseOptionsResult.argQuery = (query.type as StringLiteral).value;
+      program.stateMap(PortalCoreKeys.browse).set(target, browseOptionsResult);
+    }
+  }
+}
+
+export async function isFileExist(host: CompilerHost, filePath: string) {
+  try {
+    const stats = await host.stat(filePath);
+    return stats.isFile();
+  } catch (e: any) {
+    if (e.code === "ENOENT" || e.code === "ENOTDIR") {
+      return false;
+    }
+    throw e;
+  }
+}
+
+export function getBrowse(program: Program, target: Type) {
+  return program.stateMap(PortalCoreKeys.browse).get(target);
+}
+
+export function getBrowseArgQuery(program: Program, target: Type) {
+  return getBrowse(program, target)?.argQuery;
+}
+
+export function isARMResource(
```

Comment:
```suggestion
export function checkIsArmResource(
```
just as this is actually reporting diagnostics

Reply:
also do you want to return early if its failing here?

Reply:
fixed :) yes I want to return early :) added return statements :)